`ServiceResult` minor code improvements and documentation

myabc opened the pull request:
# Ticket

N/A

# What are you trying to accomplish?

* Add comprehensive docs for `ServiceResult` using YARD.
* Some small code improvements to the `ServiceResult` class, including fixing a `Lint/DuplicateMethods` offence.

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

# Merge checklist

- [ ] Added/updated tests
- [X] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
Fix duplicate ServiceResult#message definition

Fixes `Lint/DuplicateMethods` offense.

myabc committed:
DRY up ServiceResult.success, .failure methods

myabc committed:
Use alias keyword consistently (with barewords)

myabc committed:
Document ServiceResult with YARD

Also makes internal constants private.

NobodysNightmare commented on app/services/service_result.rb:
```diff
@@ -28,51 +28,68 @@
+  # @return [Object, nil] the Service call result object. This can also be nil.
+  attr_accessor :result
+
+  # @return [ActiveModel::Errors] errors resulting from the Service call.
```
I think there are also cases where `errors` is not an `ActiveModel::Errors`, but something simpler like a string or an array of strings<sup>[citation needed]</sup>

NobodysNightmare replied:
Additionally I'd suggest to add `nil` to the possible responses. I am very sure that we don't always add an empty errors representation. Most simple case is `ServiceResult.success(...)`

mereghost replied:
In general, the errors should always be an `ActiveModel::Errors` to ensure the same behaviour all around (and facilitate translations).

NobodysNightmare replied:
"should" is a stretchable term. A quick search from my side revealed [this beauty](https://github.com/opf/openproject/blob/741e8ab707010017ab0cd99b8125c676778a4cc4/spec/services/bulk_services/project_mappings/behaves_like_bulk_project_mapping_create_service.rb#L154), but I am sure there are more:

```ruby
      service_result = instance.call
      expect(service_result).to be_failure
      expect(service_result.errors).to eq("not found") # "not found" is neither array, nor ActiveModel::Errors
```

mereghost replied:
I think that was caused exactly by lack of documentation. `Storages` had exact this issue as the `ServiceResult`s bubbled to the UI, we had a patch up a ton.

I consider it being a `ActiveModel::Errors` the interface, everything else a bug. ;)

myabc replied:
@NobodysNightmare @mereghost agree! ~~that's a bug~~ that should be a bug :)

myabc replied:
`service_result_spec.rb` suggests we support arbitrary objects:

```rb
    let(:errors) { ["errors"] }

    it "is what has been provided" do
      instance.errors = errors

      expect(instance.errors).to eql errors
    end
```

mereghost replied:
Git knows how much time I have spent chasing down bugs caused exactly because of this. =/

NobodysNightmare replied:
For an unrelated reason I searched for the implementation of `message`. Turns out that this method is also written in a way that shows `ActiveModel::Errors` is just one possibility for `errors`, but we accept much more.

NobodysNightmare commented on app/services/service_result.rb:
```diff
@@ -28,51 +28,68 @@
+  # @api private
+  # @note Prefer using {.success} or {.failure} factory methods to calling {ServiceResult.new}.
+  #
+  # @param errors [ActiveModel::Errors, nil] errors resulting from the Service call.
```
Missing: `@param success`

myabc replied:
@NobodysNightmare that was intentional. Yard copies these input parameter docs for the `.success` and `.failure` factory methods (with the `@param (see #initialize)` syntax), but in the case of those methods `@param success` is not a valid input parameter.

(I couldn't find a way of excluding a param from being copied).

NobodysNightmare replied:
That's overall sad. I might mainly consume this documentation as inline code comment (not as parsed HTML yard doc) and now the limitations of the parser cause the inline docs to not properly document.

In the end I don't care too much about this, since this PR already provides a huge improvement over the missing docs we had beforehand. But it reminds me of a thing that the OpenStreetMap community refers to as ["tagging for the renderer"](https://wiki.openstreetmap.org/wiki/Tagging_for_the_renderer).

myabc replied:
I've updated the PR to define the subset of input parameters on `.success` and then copy to `.failure` and the constructor. Admittedly, it's a little less intuitive for readers of code, but better for consumers of the generated docs.

(I'm pretty sure we could accomplish the above with [Yard macros](https://www.rubydoc.info/gems/yard/file/docs/Tags.md#macro) - they certainly allow you to document dynamic methods defined with meta-programming. However my gut feeling is that this would be overkill for this use case.)

myabc replied:
The previous solution I had pushed (defining input params on `.success` and then adding the `@param (see .success)` tag on `#initialize`) ended up spitting out warnings:

```
[warn]: @param tag has unknown parameter name: errors 
    in file `app/services/service_result.rb' near line 79
[warn]: @param tag has unknown parameter name: message 
    in file `app/services/service_result.rb' near line 79
[warn]: @param tag has unknown parameter name: message_type 
    in file `app/services/service_result.rb' near line 79
[warn]: @param tag has unknown parameter name: state 
    in file `app/services/service_result.rb' near line 79
[warn]: @param tag has unknown parameter name: dependent_results 
    in file `app/services/service_result.rb' near line 79
[warn]: @param tag has unknown parameter name: result 
    in file `app/services/service_result.rb' near line 79
```

It turns out that, in order to define `@param`s for a method that accepts arbitrary kwargs (i.e. `**`), you need to first define an `@overload`.

After trying a couple different combinations of `@overload`, `@param (see ..) and macros, I ended up going for a macro-based solution. 

For doc consumers who read source code, it's not great, but not unintelligible either.

For those who like their docs in HTML, signatures and input parameters are now documented correctly for `.success`, `.failure`, as well as `#initialize`.

<img width="1058" alt="image" src="https://github.com/user-attachments/assets/34484964-2e1f-40bd-91f5-9ef327df5cd8" />

mereghost commented on app/services/service_result.rb:
```diff
@@ -90,53 +107,74 @@ def initialize(success: false,
+  ##
+  # Returns all {#result}s, including from dependent ServiceResults.
+  #
+  # @return [Array] all results.
```
I think you can note this as `Array<Object, nil>` to imply we have no idea what the results can be.

myabc replied:
In this case, `all_results` filters out any `nil` results, so this will be `Array<Object>`.

mereghost replied:
BTW, I wonder if we could change the implementation to `filter_map` 🤔

myabc replied:
@mereghost I was thinking the same thing:

```rb
    self_and_dependent.flat_map(&:result)
```

However it would introduce a behaviour change, as the current implementation can contain nil results from dependent ServiceResults. (it might be that that wasn't the intention, but since I am not the original author that is only me speculating)

NobodysNightmare commented on app/services/service_result.rb:
```diff
@@ -90,53 +107,74 @@ def initialize(success: false,
   ##
-  # Test whether the returned errors respond
-  # to the search key
+  # Test whether the returned errors include the error key.
```
```suggestion
  # Test whether the returned errors, including from dependent ServiceResults, include the error key.
```

NobodysNightmare commented on app/services/service_result.rb:
```diff
@@ -162,21 +205,45 @@ def add_dependent!(dependent)
   end
 
+  ##
+  # Callback to be executed if the Service call succeeds.
```
Tense of the sentence, as well as the word "Callback" imply that `ServiceResult` is some kind of promise, where the result is yet unknown, but at some point in the future the result will be known and then the callback will be executed.

But actually these methods are intended to replace `if` based control-flow statements.

My take:

```suggestion
  # Execute block argument if the Service call succeeded.
```

Similarly I'd also change the docs for `on_failure` and other following methods.

myabc replied:
You're right - "callback" implies asynchronous behaviour. Will correct both methods.

myabc commented on app/services/service_result.rb:
```diff
@@ -28,51 +28,68 @@
 
-  def initialize(success: false,
+  ##
+  # @api private
+  # @note Prefer using {.success} or {.failure} factory methods to calling {ServiceResult.new}.
```
The `@api private` is not correct - after looking at `UseServiceResultFactoryMethods` cop, it looks like we are only discouraging use of `success: true/false` **keywords** in the constructor.

```
      #   # good
      #   ServiceResult.new(success: some_value)
```

is acceptable.

myabc committed:
Fix missing doc for #initialize success: param

cbliard commented on app/services/service_result.rb:
```diff
@@ -162,21 +207,45 @@ def add_dependent!(dependent)
   end
 
+  ##
+  # Returns a new ServiceResult whose {#result} is the return value from the block.
```
It's new, but it's also a copy, so the current values of message, state, dependent_results, etc are copied along. And the copy happens only if the service result succeeded.
This should be made clearer
```suggestion
  # If the service call succeeded, returns a copy of the ServiceResult whose
  # {#result} is the return value from the block. Returns self otherwise.
```

I tried this in a `rails console`:
```
>> sr = ServiceResult.success
=>
#<ServiceResult:0x000000013fcf84c0
...
>> sr.add_dependent!(ServiceResult.success)
=>
[#<ServiceResult:0x000000015f591090
  @dependent_results=[],
  @errors=#<ActiveModel::Errors []>,
  @message=nil,
  @message_type=nil,
  @result=nil,
  @state=nil,
  @success=true>]
>> sr
=>
#<ServiceResult:0x000000013fcf84c0
 @dependent_results=
  [#<ServiceResult:0x000000015f591090
    @dependent_results=[],
    @errors=#<ActiveModel::Errors []>,
    @message=nil,
    @message_type=nil,
    @result=nil,
    @state=nil,
    @success=true>],
 @errors=#<ActiveModel::Errors []>,
 @message=nil,
 @message_type=nil,
 @result=nil,
 @state=nil,
 @success=true>
>> new_sr = sr.map { 42 }
=>
#<ServiceResult:0x000000015fdb2530
...
>> new_sr
=>
#<ServiceResult:0x000000015fdb2530
 @dependent_results=
  [#<ServiceResult:0x000000015f591090
    @dependent_results=[],
    @errors=#<ActiveModel::Errors []>,
    @message=nil,
    @message_type=nil,
    @result=nil,
    @state=nil,
    @success=true>],
 @errors=#<ActiveModel::Errors []>,
 @message=nil,
 @message_type=nil,
 @result=42,
 @state=nil,
 @success=true>
>>
```

myabc replied:
Good point!

cbliard commented on app/services/service_result.rb:
```diff
@@ -185,6 +254,8 @@ def map
     end
   end
 
+  # @return [Array<Object, nil>] the {#result} wrapped in an Array if the Service call succeeds, or an empty Array if
+  #   the Service call fails.
```
I find the `nil` disturbing. Sadly there is no way to indicate the size in yard notation.

```suggestion
  # @return [Array<Object>] the {#result} wrapped in an Array if the Service call succeeds, or an empty Array if
  #   the Service call fails.
```

myabc replied:
@cbliard `#to_a` wraps `#result`, which can return `Object` or `nil` (we don't currently enforce the presence of a result for a successful Service call[^1], although ofc in practice, result is usually non-nil on success).

After doing a bit of Googling, it looks like think there is a syntax for fixed-sized arrays/tuples, etc. BUT it looks like it's not possible to delimit multiple types ([see discussion here](https://github.com/lsegal/yard/discussions/1579)).

(I am guessing that `Array()` can be used to denote an empty Array, but haven't seen it used in the wild.)

```rb
  # @return [Array(Object), Array()] the {#result} wrapped in an Array if the Service call succeeds, or an empty Array if
```

[^1]: something to add to a list of possible enhancements?

cbliard commented on app/services/service_result.rb:
```diff
@@ -162,21 +207,45 @@ def add_dependent!(dependent)
   end
 
+  ##
+  # Iterates exactly once, passing the result to the block, if the Service call succeeds.
```
```suggestion
  # Iterates exactly once, passing the result to the block, if the Service call succeeded.
```

Past tense is already used in the above methods, so let's keep it consistent.

myabc replied:
I've fixed the tense for all of the param docs.

cbliard reviewed (approved):
Nice 👏 

I like the way `.success` and `.failure` are written now.

I have left some minor remarks.
I would also prefer if documentation lines were no longer than 80 characters. I find it easier to read wen it's shorter.

Also about style, some doc have verbs with a `s` at the end (`# Returns the ...`) and some don't (`Print messages ...`). That could be more consistent, but honestly I don't really care. That's still better than before anyway.

myabc commented:
@cbliard thanks for the helpful feedback.

> I would also prefer if documentation lines were no longer than 80 characters. I find it easier to read wen it's shorter.

Fixed (semi-automatically - I'd prefer if if we could set up Rubocop and/or code formatting to enforce this) 

> Also about style, some doc have verbs with a s at the end (# Returns the ...) and some don't (Print messages ...).

Fixed. I've changed all docs to the third-person for consistency.

cbliard commented:
> > I would also prefer if documentation lines were no longer than 80 characters. I find it easier to read wen it's shorter.
> 
> Fixed (semi-automatically - I'd prefer if if we could set up Rubocop and/or code formatting to enforce this)

Please no.
We should aim for 80 characters because it's more convenient to read, but having a rubocop complaining because of that is really a PITA. That's why we set 120 characters.

The editor can help. On VSCode, I use [rewrap](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) extension which wraps at my configured editor rulers. I have rulers at 80, 90, 100, 120, and 130 chars. Each time I ask it to wrap, it wraps to 80, then 90, then 100, etc.
I use it all the time.
